intermediate_source/ddp_tutorial.rst

```diff
@@ -265,8 +265,8 @@ either the application or the model ``forward()`` method.
         setup(rank, world_size)
 
         # setup mp_model and devices for this process
-        dev0 = (rank * 2) % world_size
-        dev1 = (rank * 2 + 1) % world_size
+        dev0 = rank * 2
```

Comment:
Copied the conversation from https://github.com/pytorch/tutorials/pull/1618:

Shen: 
Unfortunately, we cannot do this. This will make rank0 and rank 1 share the two GPUs, if there are just two GPUs. DDP and collective comm requires each process to exclusively work on GPUs, otherwise, the comm might hang.

Bo: 
The issue I got is like below:

I tried to run the code on 2 GPUs machine.
For rank0 -- dev0=0 dev1=1
For rank1 --- dev0=2 dev2=3

With that I got exception about wrong device ordinals since only cuda:0 and cuda:1 are valid. But for rank2, it ends up with cuda:2 and cuda:3

Any idea how to avoid such exception?

Reply:
Looks like only `demo_model_parallel` needs >= 4 GPUs. Maybe we can do the following?

1. skip `demo_model_parallel` when there are less than 4 GPUs
2. pass ngpus / 2 as world_size

Reply:
SG, changed as suggested